src/file_resolver.rs

```diff
@@ -96,28 +78,66 @@ impl FileResolver {
     pub fn set_file_contents(&mut self, path: &str, contents: String) {
         let pos = self.files.len();
 
-        self.files.push(Arc::from(contents));
+        let pathbuf = PathBuf::from(path);
+
+        self.files.push(ResolvedFile {
+            path: path.into(),
+            full_path: pathbuf.clone(),
+            contents: Arc::from(contents),
+            import_no: None,
+        });
 
-        self.cached_paths.insert(PathBuf::from(path), pos);
+        self.cached_paths.insert(pathbuf, pos);
     }
 
     /// Get the file contents of `file_no`th file if it exists
     pub fn get_contents_of_file_no(&self, file_no: usize) -> Option<Arc<str>> {
-        self.files.get(file_no).cloned()
+        self.files.get(file_no).map(|f| f.contents.clone())
     }
 
     /// Get file with contents. This must be a file which was previously
     /// add to the cache
     pub fn get_file_contents_and_number(&self, file: &Path) -> (Arc<str>, usize) {
         let file_no = self.cached_paths[file];
 
-        (self.files[file_no].clone(), file_no)
+        (self.files[file_no].contents.clone(), file_no)
+    }
+
+    /// Atempt to resolve a file, either from the cache or from the filesystem
```

Comment:
Could explain what a `Ok(None)` return value means and in what cases an `Err` is returned instead

Reply:
Comment added